[shopsys] automated fixes and additions of annotations for extended classes (#1344)
shopsys/shopsys: docs/introduction/framework-extensibility.md

    * separate users login credentials
    * share company attributes
    * change association from 1:1 to 1:N

## Making the static analysis understand the extended code
### Problem description
When extending framework classes, it may happen that tools for static analysis (e.g. PHPStan, PHPStorm) will not understand your code properly.
Imagine this situation:
- You have a controller that is dependent on a framework service:
```php
namespace Shopsys\ShopBundle\Controller\Front;

use Shopsys\FrameworkBundle\Model\Product\ProductFacade;

class ProductController
{
    /**
     * @var \Shopsys\FrameworkBundle\Model\Product\ProductFacade
     */
    protected $productFacade;

    /**
     * @param \Shopsys\FrameworkBundle\Model\Product\ProductFacade $productFacade
     */
    public function __construct(ProductFacade $productFacade)
    {
        $this->productFacade = $productFacade;
    }
}
```
- In your project, you extend the framework service:
```php
namespace Shopsys\ShopBundle\Model\Product;

use Shopsys\FrameworkBundle\Model\Product\ProductFacade as BaseProductFacade;

class ProductFacade extends BaseProductFacade
{
    public function myCustomAwesomeFunction()
    {

>         return 42;

Reviewer: _local towel service approves_

    }
}
```
- You register your extension in DI services configuration and thanks to that, your class is used in `ProductController` instead of the one from `FrameworkBundle`, so far so good:
```yaml
Shopsys\FrameworkBundle\Model\Product\ProductFacade: '@Shopsys\ShopBundle\Model\Product\ProductFacade'
```
**However, when you want to use your `myCustomAwesomeFunction()` in `ProductController`, the static analysis is not aware of that function.**
#### Solution
To fix this, you need to change the annotations properly:
```diff
class ProductController
{
      /**
-      * @var \Shopsys\FrameworkBundle\Model\Product\ProductFacade
+      * @var \Shopsys\ShopBundle\Model\Product\ProductFacade
       */
      protected $productFacade;

      /**
-      * @param \Shopsys\FrameworkBundle\Model\Product\ProductFacade $productFacade
+      * @param \Shopsys\ShopBundle\Model\Product\ProductFacade $productFacade
       */
      public function __construct(ProductFacade $productFacade)
      {
          $this->productFacade = $productFacade;
      }
}
```

> **Luckily, you do need to fix the annotations manually, there is the `fix-annotations-in-extended-classes` [phing target](./console-commands-for-application-management-phing-targets.md), that handles everything for you.**

Author: fixed in 6c3aa3e2a

> **Luckily, you do need to fix the annotations manually, there is the [Phing target `fix-annotations-in-extended-classes`](./console-commands-for-application-management-phing-targets.md#fix-annotations-in-extended-classes), that handles everything for you.**

Reviewer 2: Could this name be longer? :D ... It really can't be called just `fix-annotations`? :)

Author: thanks for your input, changed :wink:

- In framework, there is `ProductFacade` that has `ProductRepository` property
```php
namespace Shopsys\FrameworkBundle\Model\Product;

class ProductFacade
{
    /**
     * @var \Shopsys\FrameworkBundle\Model\Product\ProductRepository
     */
    protected $productRepository;

    /**
     * @return \Shopsys\FrameworkBundle\Model\Product\ProductRepository
     */
    public function getProductRepository()
    {
        retrun $this->productRepository;
    }
}
```
- In your project, you extend `ProductRepository` and `ProductFacade` as well.
- Then, in your extended facade, you want to access the repository (generally speaking, you want to access the parent's property that has a type that is extended in your project, or you want to access a method that returns a type that is already extended):
```php
namespace Shopsys\ShopBundle\Model\Product;

use Shopsys\FrameworkBundle\Model\Product\ProductFacade as BaseProductFacade;

class ProductFacade extends BaseProductFacade
{
    public function myCustomAwesomeFunction()
    {
        $this->productRepository; // static analysis thinks this is of type \Shopsys\FrameworkBundle\Model\Product\ProductRepository

>         $this->>getProductRepository(); // static analysis thinks this is of type \Shopsys\FrameworkBundle\Model\Product\ProductRepository

Reviewer: ```suggestion
        $this->getProductRepository(); // static analysis thinks this is of type \Shopsys\FrameworkBundle\Model\Product\ProductRepository
```

Author: fixed in acbc627e4

    }
}
```
- **Once again, static analysis is not aware of the extension.**
#### Solution

> To fix this, you do need to override the method or property, you just need to add proper `@method` and `@property` annotations to your class:

Reviewer: ```suggestion
To fix this, you don't need to override the method or property, you just need to add proper `@method` and `@property` annotations to your class:
```

Should it be like this?

Author: yes, fixed in acbc627e4

    * separate users login credentials
    * share company attributes
    * change association from 1:1 to 1:N

## Making the static analysis understand the extended code
### Problem 1
When extending framework classes, it may happen that tools for static analysis (e.g. PHPStan, PHPStorm) will not understand your code properly.
Imagine this situation:
- You have a controller that is dependent on a framework service:
```php
namespace Shopsys\ShopBundle\Controller\Front;

use Shopsys\FrameworkBundle\Model\Product\ProductFacade;

class ProductController
{
    /**
     * @var \Shopsys\FrameworkBundle\Model\Product\ProductFacade
     */
    protected $productFacade;

    /**
     * @param \Shopsys\FrameworkBundle\Model\Product\ProductFacade $productFacade
     */
    public function __construct(ProductFacade $productFacade)
    {
        $this->productFacade = $productFacade;
    }
}
```

> - In your project, you extend the framework service:

Reviewer: ```suggestion
- In your project, you extend the framework's `ProductFacade` service:
```

class ProductFacade extends BaseProductFacade
{
    public function myCustomAwesomeFunction()
    {
        $this->productRepository; // static analysis thinks this is of type \Shopsys\FrameworkBundle\Model\Product\ProductRepository
        $this->getProductRepository(); // static analysis thinks this is of type \Shopsys\FrameworkBundle\Model\Product\ProductRepository
    }
}
```
- **Once again, static analysis is not aware of the extension.**
#### Solution
To fix this, you don't need to override the method or property, you just need to add proper `@method` and `@property` annotations to your class:
```diff
namespace Shopsys\ShopBundle\Model\Product;

use Shopsys\FrameworkBundle\Model\Product\ProductFacade as BaseProductFacade;

+ /**
+  * @method \Shopsys\ShopBundle\Model\Product\ProductRepository getProductRepository()
+  * @property \Shopsys\ShopBundle\Model\Product\ProductRepository $productRepository
+  */
  class ProductFacade extends BaseProductFacade
  {
```
**Even this scenario is covered by `fix-annotations` phing target.**

### Problem 3
There is one kind of problem that is not fixed automatically and needs to be addressed manually.

> Shopsys Framework uses a kind of magic for working with extended entities (see `EntityNameResolver` class),

Reviewer: ```suggestion
Shopsys Framework uses a kind of magic for working with extended entities (see [`EntityNameResolver` class](https://github.com/shopsys/shopsys/blob/master/packages/framework/src/Component/EntityExtension/EntityNameResolver.php)),
```

> and static analysis tools are not aware of that fact. Imagine the following situation:

Reviewer: Can you please start a new line after the end of the sentence? Diffs are then nicer

Author: 0268ea6ee - I am not sure, is this what you want?

Reviewer: Nope, I meant from this
```
and static analysis tools are not aware of that fact. Imagine the following situation:
```
to this
```
and static analysis tools are not aware of that fact.
Imagine the following situation:
```

Sorry, I should be more descriptive

Author: reverted in 500097fd9 and fixed in c66ab1017
